server/map_rpc_server.go

```diff
@@ -321,6 +306,57 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 	return &trillian.SetMapLeavesResponse{MapRoot: newRoot}, nil
 }
 
+func doPreload(ctx context.Context, tx storage.MapTreeTX, hkv []merkle.HashKeyValue) error {
+	ctx, span := spanFor(ctx, "doPreload")
+	defer span.End()
+
+	readRev, err := tx.ReadRevision(ctx)
+	if err != nil {
+		return err
+	}
+	nidSet := make(map[string]bool)
+	type nodeAndID struct {
+		id   string
+		node storage.NodeID
+	}
+	c := make(chan nodeAndID, 2048)
+	var wg sync.WaitGroup
+
+	for _, i := range hkv {
+		wg.Add(1)
+		go func(k []byte) {
+			nid := storage.NewNodeIDFromHash(k)
+			sibs := (&nid).Siblings()
+			for _, sib := range sibs {
+				sibID := sib.String()
+				sib := sib
+				c <- nodeAndID{sibID, sib}
+			}
+			wg.Done()
+		}(i.HashedKey)
+	}
+
+	done := make(chan bool)
+	nids := make([]storage.NodeID, 0, len(hkv)*256)
+	go func() {
+		for nai := range c {
```

Comment:
is there a possibility that this loop ends with go routines from above writing to it after?

Reply:
No - this loop will not terminate until `c` has been closed (if `c` is still open and there's nothing to read it'll just block), and `c` cannot be closed until all the goroutines above have finished their work and called `wg.Done()` (since that unblocks the `wg.Sync()` below, after which `close(c)` will be called).

(More details [here](https://tour.golang.org/concurrency/4)